src/Database/TigerBeetle/Raw/Client.hs

```diff
@@ -74,6 +85,98 @@ data ClientState = ClientState
 
 newtype ClientHandle = ClientHandle { tvar :: TVar ClientState }
 
+data Address = Address { getAddress :: Text }
+  deriving (Eq, Show)
+
+toCString :: Address -> ((Ptr CChar, Int) -> IO a) -> IO a
+toCString = BS.useAsCStringLen . TE.encodeUtf8 . getAddress
+
+data ClientInitError
+  = Unexpected
+  | OutOfMemory
+  | AddressInvalid
+  | AddressLimitExceeded
+  | SystemResources
+  | NetworkSubsystem
+  deriving (Eq, Show)
+
+-- | Create a 'Client' initialization error from a 'TBInitStatus'.
+--
+-- Asserts that @err@ is not 'FFI.Success', throws an exception at
+-- runtime.
+toClientInitError :: TBInitStatus -> ClientInitError
+toClientInitError err = assert (err /= FFI.Success) $
+  case err of
+    FFI.Unexpected           -> Unexpected
+    FFI.OutOfMemory          -> OutOfMemory
+    FFI.AddressInvalid       -> AddressInvalid
+    FFI.AddressLimitExceeded -> AddressLimitExceeded
+    FFI.SystemResources      -> SystemResources
+    FFI.NetworkSubsystem     -> NetworkSubsystem
+    FFI.Success              -> error "toClientInitError: Success is not an error"
+
+type Client = ForeignPtr FFI.TBClient
+
+clientFinalizer :: Ptr TBClient -> IO ()
+clientFinalizer clientPtr = do
+  result <- FFI.tbClientDeinit clientPtr
+  case result of
+    FFI.ClientOk -> pure ()
+    FFI.ClientInvalid ->
+      error $ "tbClientFinalizer (invalid clientPtr): " ++ show clientPtr
+
+-- | Call @tb_client_init_echo@ and return a valid 'Client' upon success.
+--
+-- The finalizer on 'Client' will call @tb_client_deinit@.
+initClientEcho
+  :: ClusterId
+  -> Address
+  -> TBCompletionContext
+  -> FunPtr TBCompletionCallback
+  -> IO (Either ClientInitError Client)
+initClientEcho clusterId address completionCtx completionCallback = do
+  rawClientPtr <- malloc
+  finalizer <- FFI.makeClientFinalizer clientFinalizer
+  clientPtr <- newForeignPtr finalizer rawClientPtr
+  initStatus <- withForeignPtr clientPtr $ \cp -> do
+    toCString address $ \(addrPtr, addrLen) -> do
+      FFI.tbClientInitEcho
+        cp
+        clusterId
+        addrPtr
+        (fromIntegral addrLen)
+        completionCtx
+        completionCallback
+  case initStatus of
+    FFI.Success -> pure $ Right clientPtr
+    initError   -> pure . Left . toClientInitError $ initError
+
+-- | Call @tb_client_init@ and return a valid 'Client' upon success.
+--
+-- The finalizer on 'Client' will call @tb_client_deinit@.
+initClient
+  :: ClusterId
+  -> Address
+  -> TBCompletionContext
+  -> FunPtr TBCompletionCallback
+  -> IO (Either ClientInitError Client)
+initClient clusterId address completionCtx completionCallback = do
+  rawClientPtr <- malloc
+  finalizer <- FFI.makeClientFinalizer clientFinalizer
+  clientPtr <- newForeignPtr finalizer rawClientPtr
```

Comment:
Do we maybe want to use [mallocForeignPtr](https://hackage.haskell.org/package/base-4.21.0.0/docs/Foreign-ForeignPtr.html#v:mallocForeignPtr)? This will ensure that the actual ptr is freed (although perhaps deinit client does that)? Then we could use [addForeignPtrFinalizer](https://hackage.haskell.org/package/base-4.21.0.0/docs/GHC-ForeignPtr.html#v:addForeignPtrFinalizer) to add a finalizer.

Reply:
I was unsure but I think I'm coming around to `mallocForeignPtr` in order to pin the vector in memory so that the GC can't move it. We can still add our foreign finalizer with `addForeignPtrFinalizer` and it will get added (and run) after the `mallocForeignPtr` finalizer.

Reply:
@agentultra I believe the finalizer we add via `addForeignPtrFinazlier` will actually get run before the `mallocForeignPtr`, but I believe this is actually what we want, since the default finalizer on `mallocForeignPtr` just frees the ptr.